shared/src/main/scala/mlscript/TypeDefs.scala

```diff
@@ -541,5 +542,117 @@ class TypeDefs extends ConstraintSolver { self: Typer =>
+        updateVariance(negated, curVariance.flip(VarianceInfo.contra()))
+      case TypeRef(defn, targs) =>
+        val typeRefDef = ctx.tyDefs.getOrElse(defn.name, throw new Exception(s"Cannot find $defn type definition in context"))
+        targs.zip(typeRefDef.tvars).foreach { case (targ, tvar) =>
+          val tvarVariance = if (curVariance === VarianceInfo.contra())
+              typeRefDef.tvarVariance.getOrElse(tvar, VarianceInfo.bi()).flip(VarianceInfo.contra())
+            else
+              typeRefDef.tvarVariance.getOrElse(tvar, VarianceInfo.bi()).flip(VarianceInfo.co())
+          updateVariance(targ, tvarVariance)
+        }
+      case ComposedType(pol, lhs, rhs) =>
+        updateVariance(lhs, curVariance)
+        updateVariance(rhs, curVariance)
+      case TypeBounds(lb, ub) =>
+        updateVariance(lb, curVariance.flip(VarianceInfo.contra()))
+        updateVariance(ub, curVariance.flip(VarianceInfo.co()))
+      case ArrayType(inner) => fieldVarianceHelper(inner, curVariance)
+      case TupleType(fields) => fields.foreach {
+          case (_ , fieldTy) => fieldVarianceHelper(fieldTy, curVariance)
+        }
+      case FunctionType(lhs, rhs) =>
+        updateVariance(lhs, curVariance.flip(VarianceInfo.contra()))
+        updateVariance(rhs, curVariance.flip(VarianceInfo.co()))
+      case Without(base, names) => updateVariance(base, curVariance.flip(VarianceInfo.contra()))
+    }
+
+    do {
+      varianceUpdated = false;
+      tyDefs.foreach{
+        case t@TypeDef(Cls, nme, _, _, body, mthDecls, mthDefs, _, _) =>
+          updateVariance(body, VarianceInfo.bi())(ctx, t)
+          val stores = mthDecls.foreach { mthDef => 
+            val mthBody = ctx.mthEnv.getOrElse(
+              Right(Some(nme.name), mthDef.nme.name),
+              throw new Exception(s"Method ${mthDef.nme.name} does not exist in the context")
+            ).body
+            mthBody.foreach { case (_, body) => updateVariance(body, VarianceInfo.co())(ctx, t) }
+          }
+        case t@TypeDef(Trt, nme, _, _, body, mthDecls, mthDefs, _, _) =>
+          updateVariance(body, VarianceInfo.bi())(ctx, t)
+          val stores = mthDecls.foreach { mthDef => 
+            val mthBody = ctx.mthEnv.getOrElse(
+              Right(Some(nme.name), mthDef.nme.name),
+              throw new Exception(s"Method ${mthDef.nme.name} does not exist in the context")
+            ).body
+            mthBody.foreach { case (_, body) => updateVariance(body, VarianceInfo.co())(ctx, t) }
+          }
+        case t@TypeDef(Als, nme, _, _, body, mthDecls, mthDefs, _, _) =>
+          updateVariance(body, VarianceInfo.bi())(ctx, t)
+      }
+    } while (varianceUpdated)
+  }
+
+  case class VarianceInfo(isCovariant: Bool, isContravariant: Bool) {
+
+    def &&(that: VarianceInfo): VarianceInfo =
+      VarianceInfo(isCovariant && that.isCovariant, isContravariant && that.isContravariant)
+
+    def flip(that: VarianceInfo): VarianceInfo = that match {
```

Comment:
So, looking at the use you make of it, this definition is pretty redundant. Let's make it a simple `def flip: VarianceInfo` method that simply flips the `isCovariant` and `isContravariant` fields, as I originally had in mind when I suggested this method.

Then when you go through a covariant position, you just use `curVariance`. When you go through a contravariant one, you use `curVariance.flip`. And when you go through an invariant one, you use `VarianceInfo .in`.